appsec/src/extension/backtrace.c

```diff
@@ -0,0 +1,263 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+#include "backtrace.h"
+#include "configuration.h"
+#include "ddtrace.h"
+#include "logging.h"
+#include "php_compat.h"
+#include "php_objects.h"
+#include "string_helpers.h"
+
+static const unsigned int MAX_FRAMES_ALLOWED = 32;
+static const int NO_LIMIT = 0;
+static const int STACK_DEFAULT_TOP = 8;
+static const int STACK_DEFAULT_BOTTOM = 24;
+
+static zend_string *_frames_key;
+static zend_string *_language_key;
+static zend_string *_php_value;
+static zend_string *_exploit_key;
+static zend_string *_dd_stack_key;
+static zend_string *_frame_line;
+static zend_string *_frame_function;
+static zend_string *_frame_file;
+static zend_string *_frame_id;
+
+bool php_backtrace_frame_to_datadog_backtrace_frame( // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    zval *php_backtrace_frame, zval *datadog_backtrace_frame, zend_ulong index)
+{
+    if (Z_TYPE_P(php_backtrace_frame) != IS_ARRAY) {
+        return false;
+    }
+    HashTable *frame = Z_ARRVAL_P(php_backtrace_frame);
+    zval *line = zend_hash_str_find(frame, "line", sizeof("line") - 1);
+    zval *function =
+        zend_hash_str_find(frame, "function", sizeof("function") - 1);
+    zval *file = zend_hash_str_find(frame, "file", sizeof("file") - 1);
+    zval id;
+    ZVAL_LONG(&id, index);
+
+#ifdef TESTING
+    // In order to be able to test full path encoded everywhere lets set
+    // only the file name without path
+    char *file_name = strrchr(Z_STRVAL_P(file), '/');
+    Z_TRY_DELREF_P(file);
+    ZVAL_STRINGL(file, file_name + 1, strlen(file_name) - 1);
+#endif
+
+    array_init(datadog_backtrace_frame);
+    HashTable *datadog_backtrace_frame_ht = Z_ARRVAL_P(datadog_backtrace_frame);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_line, line);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_function, function);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_file, file);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_id, &id);
+
+    Z_TRY_ADDREF_P(function);
+    Z_TRY_ADDREF_P(file);
+
+    return true;
+}
+
+void php_backtrace_to_datadog_backtrace(
+    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    zval *php_backtrace, zval *datadog_backtrace)
+{
+    if (Z_TYPE_P(php_backtrace) != IS_ARRAY) {
+        return;
+    }
+
+    zend_long max_frames = get_global_DD_APPSEC_MAX_STACK_TRACE_DEPTH();
+    HashTable *php_backtrace_ht = Z_ARRVAL_P(php_backtrace);
+    unsigned int frames_on_stack = zend_array_count(php_backtrace_ht);
+
+    unsigned int top = MIN(max_frames, MAX_FRAMES_ALLOWED);
+    unsigned int bottom = 0;
+    if (frames_on_stack > MAX_FRAMES_ALLOWED && top == MAX_FRAMES_ALLOWED) {
+        top = STACK_DEFAULT_TOP;
+        bottom = STACK_DEFAULT_BOTTOM;
```

Comment:
Top and bottom frames must be calculated based on the actual `MAX_DEPTH` value provided through configuration

Reply:
Done. I was following rfc which was a bit unclear